### PR TITLE
FEAT-#3155: add 'DataFrame.nunique' implementation for OmniSci backend

### LINKED ISSUE #3155: Implement `pandas.DataFrame.nunique` at OmniSci backend
`nunique` is a distinct count in the sql notation, which is implemented in OmniSci itself, so only supporting in Modin is required.

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3155 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date 

<s>This branch is made on top of the #3150, so it's a blocker for this PR.</s>
#3150 is merged, PR is unblocked now.

### CODE REVIEW

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [inline modin/experimental/backends/omnisci/query_compiler.py:312]: In this case, OmniSci' watchdog (enabled by default) throws an exception (omnisci/omniscidb#90). Defaulting to pandas here to ensure pandas concordance.

File: modin/experimental/engines/omnisci_on_ray/frame/data.py
Comment (dchigarev) [inline modin/experimental/engines/omnisci_on_ray/frame/data.py:346]: [`AggregateExpr`](https://github.com/modin-project/modin/blob/f9c99c3817d35f7bf0d91df9aa9644545ea27517/modin/experimental/engines/omnisci_on_ray/frame/expr.py#L263) has a `distinct` parameter, which indicates whether the `count` should be distinct. We're [forward](https://github.com/modin-project/modin/pull/3158/files#diff-ec1bdeb1039dcfbc857a3d672d2844ddb9d5deb32f6413584bb2d476c481ad8fR321) this parameter from the query compiler level.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [inline modin/experimental/backends/omnisci/query_compiler.py:297]: this parameter is `sum` specific, we don't want it to be propagated into generic `sef._agg` method

File: modin/experimental/engines/omnisci_on_ray/frame/data.py
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:346]: Do you intentionally use this construction to get an error in case some new parameter is added by the front-end to force explicit handling of all `kwargs` params? I just feel like extracting `distinct` arg and passing it directly would make the code simpler to understand.

Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:346]: Yes, this was done on purpose so the error can indicate if some parameters are not processed at the level where they should be.

Explicit extraction of the `distinct` parameter looks not so good to me, because we're adding this corner case for a single operation in the general `agg` function.

**Review by ienkovich — DISMISSED** [review]

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (YarShev) [inline modin/experimental/backends/omnisci/query_compiler.py:343]: Doesn't this phrase contradict to that one above?
```
Axis to perform aggregation along. 0 is for columns when 1 is for rows.
```

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (YarShev) [inline modin/experimental/backends/omnisci/query_compiler.py:358]: This should be removed after #3161 merge.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (YarShev) [inline modin/experimental/backends/omnisci/query_compiler.py:353]: What place do you mean where `skipna` and `numeric_only` are above processed?

File: modin/experimental/engines/omnisci_on_ray/frame/data.py
Comment (YarShev) [inline modin/experimental/engines/omnisci_on_ray/frame/data.py:328]: Maybe we shouldn't add a docstring here to avoid a conflict with #2984?

Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:328]: I think that when I touch the function's signature it's my responsibility to change the doc-string as well. With changes in this commit, the `agg` doc-string from #2984 needs to be changed anyway. How do you think, should it be done here (where we have the context of changes) or in a #2984? I'd agree for any of these

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:353]: I'm referring to the [if-statement](https://github.com/modin-project/modin/pull/3158/files#diff-ec1bdeb1039dcfbc857a3d672d2844ddb9d5deb32f6413584bb2d476c481ad8fR350) above, if parameter values don't match these conditions then no action is required from our side to handle them (OmniSci backend flow satisfies the requirements of the parameters)

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1054]: > How does the test check that the method does not fallback to pandas? 

Look at the [`applier` implementation](https://github.com/modin-project/modin/pull/3158/files#diff-207cd19bf5b9d82183530af13df79e0f0d0e870b8ca91f0dc16d1c2a4ed66b03R1036). It grabs all the `UserWarning`s that was arisen during the function flow and ensures that the tested function is not the cause of them.

> we should probably keep float data for testing

Okay, let's wait then until #3187 will be merged

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:332]: No, in case of aggregating along columns the result would be:
```
  col1  col2            col1  col2
0    1     4    ====> 0    6    15   
1    2     5    ====>
2    3     6         
```
so single-row qc is produced.

In the opposite, if `axis=1` then the single-column frame is produced:
```
  col1  col2            0
0    1     4    ====> 0 5   
1    2     5    ====> 1 7
2    3     6          2 9
```

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py]: removed

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:314]: We have (#2945), no progress there though

**Comment by YarShev** [comment]
#3187 was merged.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (YarShev) [reply modin/experimental/backends/omnisci/query_compiler.py:335]: It seems to me the docstring (the description of `axis`) looks confused a bit in comparing to native pandas description ([link](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.aggregate.html)). Maybe, we should change it?

File: modin/experimental/engines/omnisci_on_ray/frame/data.py
Comment (YarShev) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:328]: Let's keep the docstring here because new parameter `**kwargs` was added.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (YarShev) [reply modin/experimental/backends/omnisci/query_compiler.py:317]: Oh, I see. @ienkovich , any progress there?

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:335]: 1. `OmnisciQueryCompiler._agg` is not the equivalent to the high-level `DataFrame.agg` method. It's actually a private OmniSci-backend specific method, which can deviate from the base operation flow.
2. The doc-string about axes show what the function does: 
- if `axis=0` then the function applied to each __column__ and so each column get reduced into a single scalar value, hence frame became a __single-row__
- if `axis=1` then the function applied to each __row__ and so it produces a __single-column__ qc.

Comment (YarShev) [reply modin/experimental/backends/omnisci/query_compiler.py:335]: > 1. `OmnisciQueryCompiler._agg` is not the equivalent to the high-level `DataFrame.agg` method. It's actually a private OmniSci-backend specific method, which can deviate from the base operation flow.
> 2. The doc-string about axes show what the function does:
> 
> * if `axis=0` then the function applied to each **column** and so each column get reduced into a single scalar value, hence frame became a **single-row**
> * if `axis=1` then the function applied to each **row** and so it produces a **single-column** qc.

Okay, I am getting convinced about the docstring's correctness but let's clarify that a bit adding some words about function applied.

Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:337]: Expanded axis parameter doc-string at the last commit

**Comment by YarShev** [comment]
@dchigarev , thanks, LGTM assuming tests pass.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (vnlitvinov) [inline modin/experimental/backends/omnisci/query_compiler.py:312]: don't we want to add a warning that we're defaulting to something here?

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (vnlitvinov) [inline modin/experimental/backends/omnisci/query_compiler.py:344]:
```suggestion
        if not kwargs.get("skipna", True) or kwargs.get("numeric_only"):
```
I don't see why we would want to compare to explicit `True`/`False` here - users could send any truthy or falsy objects, like `0` or `1` just as well.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (vnlitvinov) [inline modin/experimental/backends/omnisci/query_compiler.py:348]:
```suggestion
        kwargs.pop("skipna")
        kwargs.pop("numeric_only")
```
you don't have to specify a "default if missing" value when you're going to ignore the result :) it looks cleaner without that parameter IMHO

Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:348]: Python dictionaries raise `KeyError` when you're trying to pop non-existent key:
```python
>>> {"a": 1}.pop("b")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'b'
```

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:312]: This warning is arisen at the base class method

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:344]: Regarding to the `is False` check:
This is some kind of artifact of handling default parameters, sometimes `None` doesn't mean `False`, but the default value (that can actually be `True`). 
```python
>>> pandas.DataFrame({"a": [1, 2, None, 3]}).sum(skipna=False)
a   NaN
dtype: float64
>>> pandas.DataFrame({"a": [1, 2, None, 3]}).sum(skipna=None)
a    6.0
dtype: float64
>>> pandas.DataFrame({"a": [1, 2, None, 3]}).sum(skipna=True)
a    6.0
dtype: float64
```
I bet we should handle all of those cases at the front, created an issue for that: #3195

As for the `is True` check:
Agreed, will change this in the next commit.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (YarShev) [inline modin/experimental/backends/omnisci/query_compiler.py:345]: @dchigarev , @vnlitvinov , do you really think that `kwargs.get("skipna", True) is not None and not kwargs.get("skipna", True)` is better than `kwargs.get("skipna", True) is False`?

Comment (dchigarev) [inline modin/experimental/backends/omnisci/query_compiler.py:345]: Kwargs may not contain `skipna` because the processed function might not have this parameter. Getting this value once and then check whether it was originally in the params or not, to whether or not to pass it to the `super()` method will produce much more code than this unfancy construction of getting `skipna` value twice

Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:345]: At least it covers all of the cases.

And also, this is only a "temporary" solution until #3195 will be resolved.

File: modin/experimental/backends/omnisci/query_compiler.py
Comment (ienkovich) [reply modin/experimental/backends/omnisci/query_compiler.py:317]: Not really.

Comment (YarShev) [reply modin/experimental/backends/omnisci/query_compiler.py:317]: It would be good to put that in 0.10.1.

**Review by vnlitvinov — APPROVED** [review]
LGTM

**Review by YarShev — APPROVED** [review]

**Pull request merged**
